Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs

```diff
@@ -611,23 +611,22 @@ private void DeleteTempFolder()
 
             try
             {
-                var contentHashes = ReadContentHashesFromDisk(context);
-                _tracer.Debug(context, $"Enumerated {contentHashes.Count} entries by {stopwatch.ElapsedMilliseconds}ms.");
+                var contentHashes = ReadSnapshotFromDisk(context);
+                _tracer.Debug(context, $"Enumerated {contentHashes.Count} entries in {stopwatch.ElapsedMilliseconds}ms.");
 
-                var hashInfoPairs = new List<KeyValuePair<ContentHash, ContentFileInfo>>();
-                foreach (var grouping in contentHashes.GroupBy(hash => hash))
+                // We are using a list of classes instead of structs due to the maximum object size restriction
+                // When the contents on disk grow large, a list of structs surpasses the limit and forces OOM
+                var hashInfoPairs = new List<PayloadFromDisk<ContentFileInfo>>();
+                foreach (var grouping in contentHashes.GroupByHash())
                 {
-                    var contentFileInfo = new ContentFileInfo(Clock, grouping.First().fileInfo.Length, grouping.Count());
+                    var contentFileInfo = new ContentFileInfo(Clock, grouping.First().Payload.Length, grouping.Count());
                     contentCount++;
                     contentSize += contentFileInfo.TotalSize;
 
-                    hashInfoPairs.Add(
-                        new KeyValuePair<ContentHash, ContentFileInfo>(
-                            grouping.Key.hash,
-                            contentFileInfo));
+                    hashInfoPairs.Add(new PayloadFromDisk<ContentFileInfo>(grouping.Key, contentFileInfo));
                 }
 
-                return hashInfoPairs;
+                return new ContentDirectorySnapshot<ContentFileInfo>(hashInfoPairs);

```

Comment:
>rectorySnapshot<ContentFileInfo>(hashInfoPairs); [](start = 36, length = 48)

Note: This is double-allocating the lists and dynamically growing the lists twice.. (i.e. the list in hashInfoPairs as well as the listss in the snapshot...
Since  you are already looping here... should the snapshot accept a single payloadfromdisk at a time?